### ISSUE TITLE
Accessing `AsyncLocalStorage` within a query client extension (`$allOperations`) breaks under Vercel Edge functions/Cloudflare Workers

### ISSUE THREAD

**Issue by andyjy** [prisma/prisma#20104]
### Bug description

TL;DR: 
1. calling `.getStore()` on an `AsyncLocalStorage` (from `node:async_hooks`)..
2. ..*from within a Prisma client `query:` extension*..
3. ..(but not elsewhere, including a custom `client:` method within the same Prisma client extension) 
4. ..is always returning `undefined` for me (instead of returning the contents of the async store)
5. ..*when deployed to Vercel Edge functions (which runs on Cloudflare Workers)*
6. ...(but works locally + under Node.JS)

~

I'm using `AsyncLocalStorage` to pass request context around (current userId, etc).
I've been using this successfully within a Prisma Client extension to implement Row-level security with Postgres.

Works just fine under Node.JS - but I hit a weird bug(?) trying to make the same code work under the Vercel Edge functions runtime (which runs on top of Cloudflare Workers).

Vercel Edge / Cloudflare workers does support AsyncLocalStorage - and I've confirmed the same async storage is working in other parts of my code. But fails to load the async context store from within a Prisma client query extension.
(See illustrative code below.)

Sorry this isn't a good reproduction (yet) - and a very specific set of circumstances that might be caused by an incompatibility within Prisma, but may potentialy be a bug within Cloudflare workers' implementation of `AsyncLocalStorage`.

It's been driving me crazy trying to diagnose and I want to get a record written up here in case anyone else runs into this and knows something!

### How to reproduce

..illustrative code snippet for now (sorry - just wanted to start getting this out there); I'll debug further and update when I can, ideally with a MRE.

```typescript
import { AsyncLocalStorage } from "node:async_hooks";

globalThis.storage = new AsyncLocalStorage<{whatever: string}>;

const exampleExtension = Prisma.defineExtension((prisma) => {
  return prisma.$extends({
    client: {
      clientMethodsWork() {
        // within client methods works - value of `whatever` is retrieved from async store as expected
        console.log(globalThis.storage.getStore()?.whatever);
      }
    },
    query: {
      $allOperations(params) {
        // within query extension doesn't work when deployed to Vercel Edge functions (on Cloudflare Workers)
        // (i.e. using Edge/dataproxy client) - getStore() returns `undefined`
        // - works locally (specifically: under Next.JS simulation of Vercel Edge runtime)
        // - works when deployed under Node.JS runtime
        console.log(globalThis.storage.getStore()?.whatever);
        // ...
        // return query(args); etc
      }
    }
  });
});

// ...

const prisma = new PrismaClient().$extends(exampleExtension);

globalThis.storage.run({whatever: "some state to pass async"}, () => {
  // any method..
  prisma.exampleModel.findUnique(...);
});
```

### Expected behavior

Reading from AsyncLocalStorage should work anywhere within the async call stack, including within custom query client extensions like `$allOperations`.

### Prisma information

(I don't believe this is related to the schema or specific query usage.)

### Environment & setup

- OS: Vercel Edge Functions / Cloudflare Workers
- Database: N/A
- Node.js version: N/A

### Prisma Version

```
prisma                  : 4.16.2
@prisma/client          : 4.16.2
Current platform        : darwin-arm64
Query Engine (Node-API) : libquery-engine 4bc8b6e1b66cb932731fb1bdbbc550d1e010de81 (at node_modules/@prisma/engines/libquery_engine-darwin-arm64.dylib.node)
Migration Engine        : migration-engine-cli 4bc8b6e1b66cb932731fb1bdbbc550d1e010de81 (at node_modules/@prisma/engines/migration-engine-darwin-arm64)
Format Wasm             : @prisma/prisma-fmt-wasm 4.16.1-1.4bc8b6e1b66cb932731fb1bdbbc550d1e010de81
Default Engines Hash    : 4bc8b6e1b66cb932731fb1bdbbc550d1e010de81
Studio                  : 0.484.0
```

**Comment by janpio** [comment]
Quick side question: What are you trying to do? Use AsyncLocalStorage as a database alternative via a Client Extension that skips all Query Engine interaction? 👀

**Comment by andyjy** [comment]
> Quick side question: What are you trying to do? Use AsyncLocalStorage as a database alternative via a Client Extension that skips all Query Engine interaction? 👀

Hah - no! (That would be.. a curious strategy.)

> I'm using AsyncLocalStorage to pass request context around (current userId, etc).
> I've been using this successfully within a Prisma Client extension to implement Row-level security with Postgres.

To expand a bit: I’m using AsyncLocalContext to pass the per-request values for row-level-security instead of alternatives like using a per-request client extension - eg. `prisma.$extends(withRowLevelSecurity({userId: xx})` - due to the architecture of my controllers calling eachother without using dependency injection to pass the prisma client around. 

A (big) refactor to use this pattern instead would remove the requirement for using AsyncLocalStorage - but it’s been working great for many months (under the Node runtime) and without the API complexity of passing the client around via dependency injection.

(I’ve also updated the code snippet above to add a comment illustrating the use of `return query(args);` within `$allOperations` in case that was causing the confusion).

**Comment by andyjy** [comment]
> Quick side question: What are you trying to do? […] via a Client Extension that skips all Query Engine interaction? 👀

Fun fact on the side since you mention 😆 - separately (and nothing to do with AsyncLocalStorage or this issue) I am using another client extension that allows me to test locally using the Prisma Edge generated client while connecting to a local Postgres database (so without using Data Proxy).

And - this does in fact work by intercepting/skipping all query engine calls - and instead redirecting them (over http) to a separate lightweight proxy server in a separate Node process that executes them against the normal non-Edge Prisma client + query engine and returns raw responses. Support for interactive transactions was tricky but it’s been working great! I’ll try clean up the code and share soon in case of interest.

**Comment by janpio** [comment]
The absence of `return query()` indeed led me down that thought path. Looking forward to you write about your approach in some way. Sounds super interesting and advanced 💚 

(Now I'll be quite and let someone chime in about your _actual_ problem.)

**Comment by andyjy** [comment]
> Looking forward to you write about your approach in some way.

I've created the relevant feature request at https://github.com/prisma/prisma/issues/20112 and published my code for the extension with writeup in the README to https://github.com/andyjy/prisma-local-proxy / https://www.npmjs.com/package/prisma-local-proxy

**Comment by andyjy** [comment]
Update: I think the cause of the original issue above (`AsyncLocalStorage` not passed through to query extensions under Cloudflare Workers) is due to the use of [`createPrismaPromise()`](https://github.com/prisma/prisma/blob/99a96e884e3cfc799583e1bc7487f79753355d62/packages/client/src/runtime/core/request/createPrismaPromise.ts#L22C19-L22C38) - which is implemented as a "thenable". 

The Cloudflare Workers implementation of `AsyncLocalStorage` for the V8 engine is [apparently known to cause this issue with thenables](https://github.com/cloudflare/workerd/pull/283#issuecomment-1398575582). I've found [this reference to a possible workaround using `AsyncResource`](https://github.com/cloudflare/workerd/pull/282#issuecomment-1402921124)(?) but haven't gotten my head around what that refers to and whether it represents a possible solution here yet.

**Comment by andyjy** [comment]
Linked PR demonstrates what a fix within Prisma looks like.

To unblock my own project I'm going to proceed either with a `patch-package` patch to apply this PR for my own project, or potentially find a userland workaround that I can implement in my own code without patching Prisma (which I'm sure is possible, but I'm not sure is possible as simple and robustly as applying a patch).

As an aside: a reproduction of Cloudflare Workers' lack of support for passing async context across thenables is here:
https://github.com/andyjy/nextjs-tests/tree/als-thenables-edge
-> Works under node: https://nextjs-tests-git-als-thenables-edge-andyyoung.vercel.app/api
-> Loses context under Edge, unless `bind()` workaround applied: https://nextjs-tests-git-als-thenables-edge-andyyoung.vercel.app/api/edge

^- this repro doesn't include Prisma; if helpful at any point for e.g. an ecosystem test I could create a reproduction demonstrating the impact within Prisma query extensions.

**Comment by andyjy** [comment]
Should be fixed by https://github.com/cloudflare/workerd/pull/1665 - pending testing once deployed to production

**Comment by eliasnorrby** [comment]
I'm experiencing a similar problem locally, under Node.js, when trying to access async local storage in a `$on('query', ...)` handler. I'm trying to use [`nestjs-cls`](https://papooch.github.io/nestjs-cls/) to increment a counter for each query:

```ts
prisma.$on('query', () => {
  if (cls.isActive()) {
    const currentCount = cls.get('prismaQueryCount') ?? 0
    cls.set('prismaQueryCount', currentCount + 1)
  }
})
```
but `cls.isActive()` is always false (equivalent to `storage.getStore()` returning `undefined`). 

The feature I'm trying to implement is a database query counter per GraphQL operation, to be logged on request completion, to help troubleshoot when `findUnique` batching fails (#23313). Don't want to hijack the issue here, but if there's a better way than what I'm attempting I'd be happy to hear it 🥸

**Comment by andyjy** [comment]
Confirmed fixed by https://github.com/cloudflare/workerd/pull/1665  🎉 

@eliasnorrby this issue was specific to Cloudflare Workers' previously-incomplete support for thenables - your issue under Node.js would be separate cause.  I do observe you're using `$on` which I can imagine (total guess) might be calling the event receiver outside the async context.. or it may be related to how/where your `cls` variable is instantiated and the innards/requirements of `nestjs-cls`, with which I'm not familiar.  You may also find more luck via a Prisma Client Extension rather than the older `$on` event receiver. Finally your query debugging use case might be better handled by Prisma's Metrics feature(?).  Good luck!

**Comment by eliasnorrby** [comment]
Created #23397 to track the separate issue with `$on` handlers.

@andyjy Thank you for the suggestions 🙏  Finding this issue lead me to try client extensions as an alternative. Async local storage does work there – however, that ended up having me count prisma client operations (e.g. `users.findUnique`) instead of SQL queries, so I failed to detect batching failures using this strategy. 

I've looked into metrics as well, but as far as I'm aware, it can only give me the total number of queries issued by the client. What I'm looking for is a way to count the number of queries made as part of handling a specific request.
